libs/marinade-sdk-macro/src/lib.rs

```diff
@@ -0,0 +1,455 @@
+    let struct_name_stripped = struct_name_as_string.strip_suffix("Accounts");
+    if struct_name_stripped.is_none() {
+        panic!(
+            "Struct {} annotated with MarinadeInstructionAccounts is expected to have a name ending with 'Accounts'",
+            struct_name_as_string
+        );
+    }
+    let struct_name_stripped = struct_name_stripped.unwrap();
+    let infos_struct_name = proc_macro2::Ident::new(
+        format!("{}AccountInfos", struct_name_stripped).as_str(),
+        proc_macro2::Span::call_site(),
+    );
+    let data_struct_name = proc_macro2::Ident::new(
+        format!("{}Data", struct_name_stripped).as_str(),
+        proc_macro2::Span::call_site(),
+    );
+    (data_struct_name, infos_struct_name)
+}
+
+fn emit_struct_fields_non_pubkey(
+    struct_fields: &Vec<(Ident, AccountsFieldData)>,
+    base_pattern: &str,
+    f: fn(&AccountsFieldData) -> String,
+) -> Vec<TokenStream2> {
+    struct_fields.iter()
+        .map(|(_, props)| {
+            let adjusted_pattern = base_pattern.replace("{}", f(props).as_str());
+            let parsed_pattern: proc_macro2::TokenStream = adjusted_pattern.parse().unwrap();
+            if !props.type_is_pubkey {
+                quote!(#parsed_pattern)
+            } else {
+                TokenStream2::new()
+            }
+        }).collect::<Vec<_>>()
+}
+
+fn emit_struct_fields_at_pattern(
+    struct_fields: &Vec<(Ident, AccountsFieldData)>,
+    base_pattern: &str,
+    f: fn(&AccountsFieldData) -> &str,
+) -> Vec<TokenStream2> {
+    struct_fields
+        .iter()
+        .map(|(_, props)| {
+            let self_name = base_pattern.replace("{}", f(props));
+            let account_meta_def: proc_macro2::TokenStream = props
+                .account_meta_formatter
+                .replace("{}", self_name.as_str())
+                .parse()
+                .unwrap();
+            if props.type_is_pubkey {
+                quote!(#account_meta_def)
+            } else {
+                TokenStream2::new()
+            }
+        }).collect::<Vec<_>>()
+}
+
+/// Example of macro generation that will generate a new struct `TestAccountInfos`
+/// and all micro anchor implementations required for the instruction would work.
+///
+/// ```
+/// use marinade_sdk_macro::{MarinadeInstructionAccounts, MarinadeInstructionData};
+/// use borsh::{BorshDeserialize, BorshSerialize};
+///
+/// // TestData is reequired as this macro depends on existent of it, based on its name!
+/// #[derive(MarinadeInstructionData, BorshSerialize, BorshDeserialize)]
+/// #[discriminator([1,2,3,4,5,6,7,8])]
+/// pub struct TestData {
+///   pub lamports: u64
+/// }
+///
+/// #[derive(MarinadeInstructionAccounts)]
+/// #[ownerid(solana_program::bpf_loader::ID)]
+/// pub struct TestAccounts {
+///     #[account(mut)]
+///     pub marinade: solana_program::pubkey::Pubkey,
+/// }
+/// ```
+///
+/// ```
+/// use marinade_sdk_macro::MarinadeInstructionData;
```

Comment:
Looks like duplication of structs

Reply:
I'm sorry I'm not sure what do you mean by the "duplication of structs"?
The `use` statement needs to be part of the definition otherwise the comment code is not compiled.